src/AuthorizedSocketConnection.js

```diff
@@ -104,51 +97,80 @@ export default class AuthorizedSocketConnection<TContext, TCredentials> {
         maxSubscriptionsPerConnection: this.config
           .maxSubscriptionsPerConnection,
       });
-      this.config.socket.emit('app_error', {
+      this.emitError({
         code: 'subscribe_failed.subscription_limit',
       });
-
       acknowledge(cb);
       this.config.socket.disconnect();
       return;
     }
 
-    const subscriptionPromise = subscribe({
-      schema: this.config.schema,
-      document: parse(query),
-      variableValues: variables,
-      contextValue: {
-        ...this.config.context,
-        subscribe: async (...args) => {
-          const source = this.config.subscriber.subscribe(...args);
-          const filtered = AsyncUtils.filter(source, this.isAuthorized);
-
-          return filtered;
-        },
-      },
-    });
-
-    this.subscriptions.set(id, subscriptionPromise);
-
     let result;
     try {
-      result = await subscriptionPromise;
-    } catch (err) {
-      this.subscriptions.delete(id);
-      throw err;
-    }
+      if (this.subscriptions.has(id)) {
+        this.log('debug', 'Duplicate subscription attempted', { id });
 
-    if (result.errors != null) {
-      this.subscriptions.delete(id);
-      this.config.socket.emit('app_error', {
-        code: 'subscribe_failed.gql_error',
-        // $FlowFixMe
-        data: result.errors,
+        this.emitError({
+          code: 'invalid_id.duplicate',
+          data: id,
+        });
+
+        return;
+      }
+
+      const documentAST = parse(query);
+      const validationErrors = this.config.documentValidationRules
+        ? validate(
+            this.config.schema,
+            documentAST,
+            this.config.documentValidationRules,
+          )
+        : [];
```

Comment:
this should probably be the same as graphql-express, e.g. use the default validation rules here?